### ISSUE TITLE
Compiler does not catch invalid type conversion

### ISSUE THREAD

**Issue by DASPRiD** [rust-lang/rust#116886]
I tried this code:

```rust
pub struct NfcUid(pub [u8; 7]);

impl TryFrom<Vec<u8>> for NfcUid {
    type Error = TryFromSliceError;

    fn try_from(value: Vec<u8>) -> Result<Self, Self::Error> {
        value.try_into()
    }
}
```

I expected to see this happen: The compiler should complain that try_into() would go into a recursion, never yielding the correct type.

Instead, this happened: The code compiles and results in a stack overflow error.

### Meta

`rustc --version --verbose`:
```
rustc 1.73.0 (cc66ad468 2023-10-03)
binary: rustc
commit-hash: cc66ad468955717ab92600c770da8c1601a4ff33
commit-date: 2023-10-03
host: x86_64-unknown-linux-gnu
release: 1.73.0
LLVM version: 17.0.2
```

**Comment by asquared31415** [comment]
This is a known limitation, the compiler only detects *direct* unconditional recursion.  That is `my_func -> my_func -> my_func -> ...`.  Your example is `try_from -> try_into -> try_from -> try_into -> ...`.  There is another issue tracking this for some other cases, but I don't recall which.
